components/ucd/normal/tests/conformance_tests.rs

```diff
@@ -0,0 +1,153 @@
+fn test_decomposition_type_conformance() {
+    let mut passed_num: u32 = 0;
+    let mut failures: Vec<Fail> = Vec::new();
+    let mut to_consider: Set<char> = iter_all_chars().collect();
+
+    for (line_idx, line) in DT_TEST_DATA.lines().enumerate() {
+        // Remove comments and trailing whitespace
+        let line = if let Some(idx) = line.find('#') {
+            &line[0..idx]
+        } else {
+            line
+        }.trim();
+
+        // Ignore empty lines
+        if line.is_empty() {
+            continue;
+        }
+
+        // State setting lines
+        if line.starts_with('@') {
+            panic!("State setting line in DecompositionTypeTest.txt; expected none");
+        }
+
+        // Data lines
+        {
+            let fields: Vec<_> = line.split(';').map(str::trim).collect();
+            let input_range = {
+                let range_text = fields[0];
+                let mut range = range_text.split("..");
+                let low = u32::from_str_radix(range.next().unwrap(), 16).unwrap();
+                let high = range
+                    .next()
+                    .map(|c| u32::from_str_radix(c, 16).unwrap())
+                    .unwrap_or(low);
+                low..(high + 1)
+            };
+            let exp_dt = Some(get_dt_from_name(fields[1]));
+
+            for codepoint in input_range {
+                // TODO: We should be safe to use from_u32_unsafe
+                if let Some(char) = char::from_u32(codepoint) {
+                    let actual_dt = DT::of(char);
+                    if actual_dt != exp_dt {
+                        failures.push(Fail {
+                            line_num: Some(line_idx + 1),
+                            char,
+                            exp_dt,
+                            actual_dt,
+                        });
+                    } else {
+                        passed_num += 1;
+                    }
+                    to_consider.remove(&char);
+                }
+            }
+        }
+    }
+
+    // Defaults
+    for char in to_consider {
+        let exp_dt = None;
+        let actual_dt = DT::of(char);
+        if actual_dt != exp_dt {
+            failures.push(Fail {
+                line_num: None,
+                char,
+                exp_dt,
+                actual_dt,
+            });
+        } else {
+            passed_num += 1;
+        }
+    }
+
+    // Report
+    if !failures.is_empty() {
+        // TODO: Show a list of failed cases when the number is less than 1K
+        panic!(
+            "{} test cases failed! ({} passed) {{\n\
+             \n\
+             0: {:?}\n\
+             \n\
+             ...\n\
+             \n\
+             {}: {:?}\n\
+             \n\
+             }}",
+            failures.len(),
+            passed_num,
+            failures[0],
+            failures.len() - 1,
+            failures[failures.len() - 1],
+        );
+    }
+}
+
+fn get_dt_from_name(name: &str) -> DT {
```

Comment:
Should I add a test for this fn?

Also,
- Should the short forms be included? I don't think they're used in the source txt (but I haven't read the whole thing).
- Should this be part of the enum instead of a helper here?